### PR TITLE
Fix #84: allow brokers to be added to/removed from the cluster

### LINKED ISSUE #84: Support add/remove broker to existing cluster
As an author of integration tests for Kroxylicious, I am interested in testing use-cases where the number of brokers in the cluster is changed at runtime, i.e. brokers added or broker removed.   I want to be able to control this programmatically from within the test.

I am imagining an extension to `io.kroxylicious.testing.kafka.api.KafkaCluster` interface that lets me:

`#addBroker() => nodeId`
`#removeBroker(int nodeId)`

I want the feature to work in both kraft and zookeeper mode.  In kraft mode, I only need the ability to add a broker in the role of `broker`.  It would be reasonable if the API rejected attempts to remove the node that is the `controller`.

I don't need the `#removeBroker` API to do any partition reassignment (to evacuate the broker).

### PR DESCRIPTION
**Pull request by k-wall**
Extends the `KafkaCluster` API to introduce the following functionality:

* `addBroker`
* `removeBroker`

The methods are blocking so won't return until the broker is added or removed.  I decided my original idea of returning Futures didn't really fit in the existing API or fit well with likely use-cases (the test would probably just block until the Future completed).

### CODE REVIEW

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 1]
```diff
@@ -246,20 +266,20 @@ public String getKafkaVersion() {
         return kafkaImage.getVersionPart();
     }
 
-    private Stream<GenericContainer<?>> allContainers() {
+    private synchronized Stream<GenericContainer<?>> allContainers() {
         return Stream.concat(
-                this.brokers.stream(),
+                this.brokers.values().stream(),
                 Stream.ofNullable(this.zookeeper));
     }
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:279]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:279]: @sonatype-lift ignore

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 2]
```diff
@@ -112,7 +129,7 @@ public TestcontainersKafkaCluster(DockerImageName kafkaImage, DockerImageName zo
         this.zookeeperImage = Optional.ofNullable(zookeeperImage).orElse(DEFAULT_ZOOKEEPER_IMAGE);
         this.clusterConfig = clusterConfig;
 
-        var name = Optional.ofNullable(clusterConfig.getTestInfo())
+        this.name = Optional.ofNullable(clusterConfig.getTestInfo())
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:132]: @sonatype-lift ignore

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 3]
```diff
@@ -131,14 +157,14 @@ private KafkaConfig buildBrokerConfig(KafkaClusterConfig.ConfigHolder c, Path te
...
-            anonPorts = preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum());
-            interBrokerPorts = preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum());
-            controllerPorts = allocateControllerPorts(clusterConfig, preallocator);
+            Utils.putAllListEntriesIntoMapKeyedByIndex(preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum()), externalPorts);
+            Utils.putAllListEntriesIntoMapKeyedByIndex(preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum()), anonPorts);
+            Utils.putAllListEntriesIntoMapKeyedByIndex(preallocator.preAllocateListeningSockets(clusterConfig.getBrokersNum()), interBrokerPorts);
+            Utils.putAllListEntriesIntoMapKeyedByIndex(allocateControllerPorts(preallocator, clusterConfig), controllerPorts);
         }
 
         kafkaEndpoints = new KafkaClusterConfig.KafkaEndpoints() {
```
Comment (k-wall) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:170]: I feel that `KafkaEndpoints` needs refactoring into something better.   I wondering between:

* turning it into a concrete class and give it the responsibility/ownership for the externalPorts/anonPorts fields
* or, having TestcontainersKafkaCluster|InVMKafkaCluster implement the KafkaEndpoints interface themselves.

I'm inclined to make that change in a separate PR.

File: api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java [hunk 4]
```diff
@@ -22,6 +22,25 @@ public interface KafkaCluster extends AutoCloseable {
      */
     void start();
 
+    /**
+     * adds a new broker to the cluster. Once this method returns the caller is guaranteed that the new broker will
+     * be incorporated into the cluster.  In kraft mode, the broker added will always from the role 'broker'.
```
Comment (k-wall) [inline api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java:27]:
```suggestion
     * be incorporated into the cluster.  In kraft mode, the broker added will always be in the role 'broker'.
```

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:280]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:280]: @sonatype-lift ignore

File: impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java [hunk 5]
```diff
@@ -40,6 +50,64 @@ public class Utils {
...
+     * Reassign all kafka internal topic partitions that exist have a replica on <code>fromNodeId</code> and don't
+     * have at least one replica elsewhere in the cluster.
+     *
+     * @param connectionConfig the connection config
+     * @param fromNodeId nodeId being evacuated
+     * @param toNodeIs replacement nodeId
+     * @param timeout the timeout
+     * @param timeUnit the time unit
+     */
+    public static void awaitReassignmentOfKafkaInternalTopicsIfNecessary(Map<String, Object> connectionConfig, int fromNodeId, int toNodeIs, int timeout,
```
Comment (franvila) [inline impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:63]: Typo error in the `toNodeIs` parameter, it should be `toNodeId`

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java [hunk 6]
```diff
@@ -69,6 +71,49 @@ public void kafkaClusterZookeeperMode() throws Exception {
         }
     }
 
+    @Test
+    public void kafkaClusterZookeeperAddBroker() throws Exception {
+        int brokersNum = 1;
+        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
+                .testInfo(testInfo)
+                .brokersNum(brokersNum)
+                .kraftMode(true)
```
Comment (franvila) [inline impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:80]: There is a mismatch between the name of the test using `Zookeeper` and the configuration of the cluster `kraftMode(true)` that should be `kraftMode(false)`

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java [hunk 7]
```diff
@@ -69,6 +71,49 @@ public void kafkaClusterZookeeperMode() throws Exception {
...
+        }
+    }
+
+    @Test
+    public void kafkaClusterZookeeperRemoveBroker() throws Exception {
+        int brokersNum = 2;
+        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
+                .testInfo(testInfo)
+                .brokersNum(brokersNum)
+                .kraftMode(true)
```
Comment (franvila) [inline impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:102]: There is a mismatch between the name of the test using `Zookeeper` and the configuration of the cluster `kraftMode(true)` that should be `kraftMode(false)`

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java [hunk 8]
```diff
@@ -69,6 +71,49 @@ public void kafkaClusterZookeeperMode() throws Exception {
...
+            int nodeId = cluster.addBroker();
+            assertThat(nodeId).isEqualTo(1);
+            assertThat(cluster.getNumOfBrokers()).isEqualTo(brokersNum + 1);
+            assertThat(cluster.getBootstrapServers().split(",")).hasSize(brokersNum + 1);
+            verifyRecordRoundTrip(brokersNum + 1, cluster);
+        }
+    }
+
+    @Test
+    public void kafkaClusterZookeeperRemoveBroker() throws Exception {
```
Comment (franvila) [inline impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:97]: Shouldn't be necessary to create another test to remove a broker in kraftMode or do you consider that it is fine testing only for zookeeper?

**Review by franvila — APPROVED** [review]
LGTM, but it needs some small changes in the tests. Thanks @k-wall!!

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 7 above)
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java]: Done.

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 8 above)
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:97]: I added a new test to cover the case.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java (hunk 3 above)
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:170]: I opened a separate PR for this #109

**Review by SamBarker — APPROVED** [review]
Whoops I left most of my feedback on https://github.com/kroxylicious/kroxylicious-junit5-extension/pull/109... 

This also LGTM.

Thanks @k-wall
